### PR TITLE
[Native] Add flow control configuration for http client

### PR DESCRIPTION
**Pull request by kewang1024**
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-native-execution/presto_cpp/main/common/Configs.cpp [hunk 1]
```diff
@@ -210,6 +210,18 @@ uint64_t SystemConfig::httpMaxAllocateBytes() const {
...
+}
+
+folly::Optional<uint64_t> SystemConfig::httpReceiveStreamWindowSize() const {
+  return optionalProperty<uint64_t>(std::string(kHttpMaxAllocateBytes));
+}
+
+folly::Optional<uint64_t> SystemConfig::httpReceiveSessionWindowSize() const {
+  return optionalProperty<uint64_t>(std::string(kHttpMaxAllocateBytes));
+}
+
```
Comment (tanjialiang) [inline presto-native-execution/presto_cpp/main/common/Configs.cpp:224]: We might need some detailed documentations on these parameters.

File: presto-native-execution/presto_cpp/main/http/HttpClient.cpp [hunk 2]
```diff
@@ -246,6 +246,21 @@ class ConnectionHandler : public proxygen::HTTPConnector::Callback {
...
+        SystemConfig::instance()->httpReceiveSessionWindowSize();
+    if (httpInitialReceiveWindow.hasValue() &&
+        httpReceiveStreamWindowSize.hasValue() &&
+        httpReceiveSessionWindowSize.hasValue()) {
+      session->setFlowControl(
+          httpInitialReceiveWindow.value(),
+          httpReceiveStreamWindowSize.value(),
+          httpReceiveSessionWindowSize.value());
+    }
+
```
Comment (tanjialiang) [inline presto-native-execution/presto_cpp/main/http/HttpClient.cpp:263]: Can we store them as static const so fetching is a one time thing? @xiaoxmeng

File: presto-native-execution/presto_cpp/main/common/Configs.h [hunk 3]
```diff
@@ -132,6 +132,19 @@ class SystemConfig : public ConfigBase {
   /// the received http response data.
   static constexpr std::string_view kHttpMaxAllocateBytes{
       "http-server.max-response-allocate-bytes"};
+  // Http client size of initial receive window size in bytes for all ingress
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/common/Configs.h:135]:
```
The http client advertised per-connection window size set on the established connection, only ...
```
The same for the other two configs.

File: presto-native-execution/presto_cpp/main/http/HttpClient.cpp [hunk 4]
```diff
@@ -246,6 +246,21 @@ class ConnectionHandler : public proxygen::HTTPConnector::Callback {
   }
 
   void connectSuccess(proxygen::HTTPUpstreamSession* session) override {
+    auto httpInitialReceiveWindow =
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/http/HttpClient.cpp:249]: Shall we get these configs on http client constructor and pass to ConnectionHandler on constructor? Then we can avoid accessing the system config in an async execution env? Also could you help also check if we can only set some of the three parameters but not all? Thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@kewang1024 thanks for adding this. It should be helpful!

**Comment by tanjialiang** [comment]
Thanks @kewang1024 . I think if our current http version does not support the first 2 parameters then we probably want to just remove them. We can add them back when we need them. I'm also thinking making this run-time configurable. Does it make sense, as a follow-up, to add to operational endpoint to allow changing of this parameter at runtime?

**Comment by kewang1024** [comment]
Yeah, sounds good, let's revisit the PR when needed, close the PR for now

**Comment by tanjialiang** [comment]
> Yeah, sounds good, let's revisit the PR when needed, close the PR for now

Are we not adding the third one? Or is that one currently useless for our http version as well?

**Pull request closed without merging**